tabular/src/autogluon/tabular/models/xgboost/xgboost_utils.py

```diff
@@ -30,66 +29,27 @@ def convert_ag_metric_to_xgbm(ag_metric_name, problem_type):
     return _ag_to_xgbm_metric_dict.get(problem_type, dict()).get(ag_metric_name, None)
 
 
-def _softmax(x):
-    return np.exp(x) / np.sum(np.exp(x), axis=1, keepdims=True)
-
-
-def _sigmoid(x):
-    return 1.0 / (1.0 + np.exp(-x))
-
-
-def func_generator(metric, is_higher_better, needs_pred_proba, problem_type):
+def func_generator(metric, problem_type: str):
+    """Create a custom metric compatible with XGBoost, based on the XGBoost 1.6+ API"""
+    sign = -1 if metric.greater_is_better else 1
+    needs_pred_proba = not metric.needs_pred
     if needs_pred_proba:
-        if problem_type == MULTICLASS:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                y_hat = _softmax(y_hat)
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
-        elif problem_type == SOFTCLASS:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                y_hat = _softmax(y_hat)
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
-        elif problem_type == BINARY:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                y_hat = _sigmoid(y_hat)
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
-        else:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
+        def custom_metric(y_true, y_hat):
+            return sign * metric(y_true, y_hat)
     else:
-        if problem_type == MULTICLASS:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                y_hat = y_hat.argmax(axis=1)
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
-        elif problem_type == SOFTCLASS:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
+        if problem_type in [MULTICLASS, SOFTCLASS]:
+            def custom_metric(y_true, y_hat):
                 y_hat = y_hat.argmax(axis=1)
-                res = metric(y_true, y_hat)
-                return metric.name, -1 * res if is_higher_better else res
+                return sign * metric(y_true, y_hat)
         elif problem_type == BINARY:
-            def function_template(y_hat, data):
-                y_true = data.get_label()
-                y_hat = _sigmoid(y_hat)
+            def custom_metric(y_true, y_hat):
                 y_hat = np.round(y_hat)
```

Comment:
Not sure if this is a big deal, but this might have some float point issues:
```python
In [5]: {i: np.around(i) for i in np.arange(0, 1, 0.1)}
Out[5]:
{0.0: 0.0,
 0.1: 0.0,
 0.2: 0.0,
 0.30000000000000004: 0.0,
 0.4: 0.0,
 0.5: 0.0,  <<< should be 1.0
 0.6000000000000001: 1.0,
 0.7000000000000001: 1.0,
 0.8: 1.0,
 0.9: 1.0}
```

there's a way to get it done and it is actually faster:
```python
In [15]: %%timeit
    ...: {i: int(i>=0.5) for i in np.arange(0, 1, 0.1)}
    ...:
    ...:
4.89 µs ± 8.45 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
vs 
```python
In [17]: %%timeit
    ...: {i: np.around(i) for i in np.arange(0, 1, 0.1)}
    ...:
    ...:
36.2 µs ± 818 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
Thanks for taking a look! I don't want to mess with this code any more prior to release, but you can send a PR for post-release follow-up